packages/nx/src/command-line/import/import.ts

```diff
@@ -0,0 +1,258 @@
+import {
+  getPackagesInPackageManagerWorkspace,
+  needsInstall,
+} from './utils/needs-install';
+
+const rmAsync = promisify(rm);
+const mkdirAsync = promisify(mkdir);
+const readdirAsync = promisify(readdir);
+
+const importRemoteName = '__tmp_nx_import__';
+
+export interface ImportOptions {
+  /**
+   * The remote URL of the repository to import
+   */
+  sourceRemoteUrl: string;
+  /**
+   * The branch or reference to import
+   */
+  ref: string;
+  /**
+   * The directory in the source repo to import
+   */
+  source: string;
+  /**
+   * The directory in the destination repo to import into
+   */
+  destination: string;
+
+  verbose: boolean;
+  interactive: boolean;
+}
+
+export async function importHandler(options: ImportOptions) {
+  let { sourceRemoteUrl, ref, source, destination } = options;
+
+  output.log({
+    title:
+      'Nx will walk you through the process of importing code from another repository into this workspace:',
+    bodyLines: [
+      `1. Nx will clone the other repository into a temporary directory`,
+      `2. Code to be imported will be moved to the same directory it will be imported into on a temporary branch`,
+      `3. The code will be merged into the current branch in this workspace`,
+      `4. Nx will recommend plugins to integrate tools used in the imported code with Nx`,
+      `5. The code will be successfully imported into this workspace`,
+      '',
+      `Git history will be preserved during this process`,
+    ],
+  });
+
+  const tempImportDirectory = join(tmpdir, 'nx-import');
+
+  if (!sourceRemoteUrl) {
+    sourceRemoteUrl = (
+      await prompt<{ sourceRemoteUrl: string }>([
+        {
+          type: 'input',
+          name: 'sourceRemoteUrl',
+          message:
+            'What is the Remote URL of the repository you want to import?',
+          required: true,
+        },
+      ])
+    ).sourceRemoteUrl;
+  }
+
+  const sourceRepoPath = join(tempImportDirectory, 'repo');
+  const spinner = createSpinner(
+    `Cloning ${sourceRemoteUrl} into a temporary directory: ${sourceRepoPath}`
+  ).start();
+  try {
+    await rmAsync(tempImportDirectory, { recursive: true });
+  } catch {}
+  await mkdirAsync(tempImportDirectory, { recursive: true });
+
+  const sourceGitClient = await cloneFromUpstream(
+    sourceRemoteUrl,
+    sourceRepoPath
+  );
+  spinner.succeed(`Cloned into ${tempImportDirectory}`);
+
+  if (!ref) {
+    const branchChoices = await sourceGitClient.listBranches();
+    ref = (
+      await prompt<{ ref: string }>([
+        {
+          type: 'autocomplete',
+          name: 'ref',
+          message: `Which branch do you want to import?`,
+          choices: branchChoices,
+          /**
+           * Limit the number of choices so that it fits on screen
+           */
+          limit: process.stdout.rows - 3,
+          required: true,
+        } as any,
+      ])
+    ).ref;
+  }
+
+  if (!source) {
+    source = (
```

Comment:
Validate this exists.

Q: Can we valdiate the project?

Reply:
Validated